File: 프로그래머스/l3_이중우선순위큐.py
import heapq

def solution(operations):
    heap = []

    for operation in operations:
        o, n = operation.split()

        if o =='I':
            heapq.heappush(heap, int(n))

        elif o =='D' and heap:
            if n =='1':
                heap.remove(max(heap))
                heapq.heapify(heap)
            elif n=='-1':
                heapq.heappop(heap)
    print(heap)

    if not heap:
        return [0,0]
    else:
        return [max(heap), min(heap)]


solution(["I 16","D 1"])

solution(["I 7","I 5","I -5","D -1"])











# import heapq

# def solution(oper):
#     queue = []
#     for op in oper:
#         o, n = op.split()
        
#         if o == 'I':
#             heapq.heappush(queue, int(n))
#         elif o =='D' and queue:
#             if n=='1':
#                 queue.remove(max(queue))
#                 heapq.heapify(queue)
#             elif n=='-1':
#                 a= heapq.heappop(queue)
#     if not queue:
#         return [0,0]
#     else:
#         return [max(queue), min(queue)]

File: 프로그래머스/JAVA/l3_이중우선순위큐.java
import java.util.Arrays;
import java.util.Collections;
import java.util.PriorityQueue;

public class l3_이중우선순위큐 {
    public static void main(String[] args){

        String[] operations = {"I 7","I 5","I -5","D -1"};
        int[] answer = solution(operations);
        System.out.println(Arrays.toString(answer));

    }

    public static int[] solution(String[] operations) {
        PriorityQueue<Integer> minHeap = new PriorityQueue<>();
        PriorityQueue<Integer> maxHeap = new PriorityQueue<>(Collections.reverseOrder());

        for(String operation:operations){
            String input[] = operation.split(" ");
            switch(input[0]) {
                case "I":
                    minHeap.add(Integer.parseInt(input[1]));
                    maxHeap.add(Integer.parseInt(input[1]));
                    break;
                case "D":
                    if (minHeap.size()>0){
                        if(Integer.parseInt(input[1]) ==1){
                            // 최대값 삭제
                            int max = maxHeap.poll();
                            minHeap.remove(max);
                        }else{
                            // 최소값 삭제
                            int min = minHeap.poll();
                            maxHeap.remove(min);
                        }
                    }
                    break;
            }
        }
        int[] answer = {0,0};

        if(minHeap.size()>=2){
            answer[0] = maxHeap.poll();
            answer[1] = minHeap.poll();
        }
        System.out.println(Arrays.toString(answer));
        return answer;

    }
}
